File: backend/routes/film.go
package routes

import (
	"FilmFindr/controller"
	"FilmFindr/helpers"
	"FilmFindr/middleware"
	"FilmFindr/service"

	"github.com/gin-gonic/gin"
)

func FilmRoute(server *gin.Engine, filmController controller.FilmController, jwtService service.JWTService) {
	film := server.Group("/api/v1/films")
	{
		film.GET("/", filmController.GetAllFilm)
		film.GET("/:id", filmController.GetFilmById)
		film.GET("/get-top-film", filmController.GetTopFilm)
		film.GET("/get-trending-film", filmController.GetTrendingFilm)
		film.GET("/search", filmController.SearchFilm)
		film.POST("/", middleware.Authenticate(jwtService), middleware.AuthorizeRole(helpers.ENUM_ROLE_ADMIN), filmController.CreateFilm)
		film.PUT("/:id", middleware.Authenticate(jwtService), middleware.AuthorizeRole(helpers.ENUM_ROLE_ADMIN), filmController.UpdateFilm)
		film.PATCH("/:id/status", middleware.Authenticate(jwtService), middleware.AuthorizeRole(helpers.ENUM_ROLE_ADMIN), filmController.UpdateStatus)
		film.DELETE("/:id", middleware.Authenticate(jwtService), middleware.AuthorizeRole(helpers.ENUM_ROLE_ADMIN), filmController.DeleteFilm)
		film.POST("/add-film-genre", middleware.Authenticate(jwtService), middleware.AuthorizeRole(helpers.ENUM_ROLE_ADMIN), filmController.AddFilmGenre)
		film.DELETE("/delete-film-genre", middleware.Authenticate(jwtService), middleware.AuthorizeRole(helpers.ENUM_ROLE_ADMIN), filmController.DeleteFilmGenre)
	}
}

File: frontend/src/service/film.js
import axios from "axios";
import { BASE_API_URL } from "../utils/constant";

const BASE_URL = `${BASE_API_URL}/films`;
export const getAllFilm = async (page) => {
    const response = await axios.get(`${BASE_URL}/?page=${page}`);
    return response.data;
};

export const getFilmById = async (id) => {
    const response = await axios.get(`${BASE_URL}/${id}`);
    return response.data;
};

export const getTopFilm = async () => {
    const response = await axios.get(`${BASE_URL}/get-top-film`);
    return response.data;
};

export const getTrendingFilm = async () => {
    const response = await axios.get(`${BASE_URL}/get-trending-film`);
    return response.data;
};

export const searchFilm = async (keyword) => {
    const response = await axios.get(`${BASE_URL}/search?keyword=${keyword}`);
    return response;
}

export const createFilm = async (data) => {
    const response = await axios.post(`${BASE_URL}/`, data, {
        withCredentials: true,
    });
    return response;
};

export const updateFilm = async (id, data) => {
    const response = await axios.put(`${BASE_URL}/${id}`, data, {
        withCredentials: true,
    });
    return response;
};

export const updateFilmStatus = async (id, data) => {
    const response = await axios.patch(`${BASE_URL}/${id}/status`, data, {
        withCredentials: true,
    });
    return response;
};

export const deleteFilm = async (id) => {
    const response = await axios.delete(`${BASE_URL}/${id}`, {
        withCredentials: true,
    });
    return response;
};

export const addFilmGenre = async (data) => {
    const response = await axios.post(`${BASE_URL}/add-film-genre`, data, {
        withCredentials: true,
    });
    return response;
};

export const deleteFilmGenre = async (data) => {
    const response = await axios.delete(`${BASE_URL}/delete-film-genre`, data, {
        withCredentials: true,
    });
    return response;
};
